### ISSUE TITLE
[FR] support for `zref` and `zref-clever`

### ISSUE THREAD

**Issue by gusbrs** [T-F-S/tcolorbox#206]
Hi, I'm author of the package `zref-clever` and I'd love to see `tcolorbox` support `zref` and `zref-clever`, as it does the standard referencing system and `cleveref`.

I'm not specially acquainted with `tcolorbox` but I did take a look at the package and I think the requirements for that should not be complicated, since it basically uses `\refstepcounter` systematically (as far as I could see), which simplifies things.

As far as I can tell, to support `zref` it is enough to provide the label options for `\zlabel`, as is done for the standard `\label`.  Something like this should be sufficient:

```tex
\tcbset{%
  zlabel/.style={phantom={\zlabel{#1}}},%
  step and zlabel/.style 2 args={phantom={\refstepcounter{#1}\zlabel{#2}}},%
  phantomzlabel/.style={phantom={\ifdefined\phantomsection\phantomsection\fi\zlabel{#1}}},%
}
```

`\zlabel` requires the `zref-user` `zref` module.

So that we could have a document like:

```tex
\documentclass{article}

\usepackage{tcolorbox}
\usepackage{zref-user}
\usepackage{zref-titleref}
\usepackage{nameref}

\tcbset{%
  zlabel/.style={phantom={\zlabel{#1}}},%
  step and zlabel/.style 2 args={phantom={\refstepcounter{#1}\zlabel{#2}}},%
  phantomzlabel/.style={phantom={\ifdefined\phantomsection\phantomsection\fi\zlabel{#1}}},%
}

\newcounter{mycounter}
\setcounter{mycounter}{4}
\newtcolorbox[auto counter,number within=section]{pabox}[2][]{%
  colback=red!5!white,colframe=red!75!black,fonttitle=\bfseries,
  title=Examp.~\thetcbcounter: #2,#1}

\begin{document}

\section{Foo bar}
\zlabel{sec:section-1}

\begin{pabox}[zlabel={myautocounter}]{Title with number}
  This box is automatically numbered with \zref{myautocounter} on page
  \zpageref{myautocounter}. Inside the box, the \thetcbcounter\ can
  also be referenced by |\thetcbcounter|.
  The real counter name is \texttt{\tcbcounter}.
\end{pabox}

\begin{pabox}[step and zlabel={mycounter}{steppedcounter}]{Title with number}
  This box is automatically numbered with \zref{steppedcounter} on page
  \zpageref{steppedcounter}. Inside the box, the \thetcbcounter\ can also be
  referenced by |\thetcbcounter|.  The real counter name is
  \texttt{\tcbcounter}.
\end{pabox}

\begin{tcolorbox}[phantomzlabel={phantomlabel}]
  This is another \textbf{tcolorbox}, \zref{phantomlabel} on page
  \zpageref{phantomlabel}.  \tcblower Here, you see the lower part of the box.
\end{tcolorbox}

\begin{pabox}[nameref={Title or anything else},zlabel={mynamelabel}]{Title text}
  This is a tcolorbox.
\end{pabox}
This box is automatically numbered with \zref{mynamelabel} on page
\zpageref{mynamelabel}.  The box is titled \zref[title]{mynamelabel}.

\end{document}
```

The only thing not working there is the `nameref` option, which would normally be handled automatically by the macros which set `\@currentlabelname`, but in this case `tcolorbox` is setting it manually (as far as I understood).  I'm not sure of the best approach here.  The `zref-titleref` incantation corresponding to `namref`'s `\NR@gettitle` is `\ztitlerefsetup{title=...}`.  Just adding it to the existing `nameref` option would need some checks, of course, since the option needs to work regardless of the presence of `zref-titleref`.  Perhaps a separate option like:

```tex
\tcbset{%
  ztitleref/.style={phantom={\ztitlerefsetup{title={#1}}}},%
}
```
would get things working.  But again, I'm not sure of the best approach. It seems a little odd to have two options that essentially do the same thing. Hence, a better alternative may be:

```tex
\tcbset{%
  nameref/.code={\tcb@gettitle{#1}%
    \ifdefined\ztitlerefsetup\ztitlerefsetup{title=\@currentlabelname}\fi},%
}
```
This would also ensure that whatever sanitizing `nameref` has done to `\@currentlabelname` also reaches `zref-titleref`. `zref-titleref` can also sanitize the title, but not using `\ztitlerefsetup` (like that, without further options). Alternatively, we could use: 

```tex
\tcbset{%
  nameref/.code={\tcb@gettitle{#1}%
    \ifdefined\zref@titleref@setcurrent\zref@titleref@setcurrent{#1}\fi},%
}
```
So that `zref-titleref` handle sanitizing on its side. I think this might be preferable since, technically, `zref-titleref` does not need `nameref` to work and, in its absence, `tcolorbox` just passes the content along.

`\ztitlerefsetup` and `\zref@titleref@setcurrent` require `zref-titleref` `zref` module.

As to `zref-clever`, it provides functionality similar to that of `cleveref` but based on `zref`.  So, some of the support `tcolorbox` provides for `cleveref` would also be meaningful.  The most important thing would be `label type`, and a corresponding option could be provided like:

```tex
\tcbset{%
  zlabel type/.code={\zcsetup{reftype={#1}}},%
}
```

`\zcsetup` requires `zref-clever`.

Besides that, there are also `crefname` and `Crefname`.  Here I think just transposing things might not be the best approach because, even if there is equivalent functionality, the UI is quite different and based on a keyval system.  If I get it correctly, the main purpose of these options (from the perspective of `tcolorbox`) is to spare the user from the complexity of using directly the internal counter generated by the `auto counter` option.  If that's indeed the case, possibly the most sensible thing to do is to set `zref-clever`'s `countertype` option, and that'd be enough to "plug it in" for `zref-clever` and the user could set the reference type at will (things which in `cleveref` would require `\crefformat`).  So, something like:

```tex
\tcbset{
  new/.cd,
  zref-clever~countertype/.code={
    \exp_args:Ne\zcsetup{countertype={tcb@cnt@\tcb@new@boxname={#1}}}
  },
}
```
Would provide all the needed handles for the user.

All in all, we could get a document like:

```tex
\documentclass{article}

\usepackage{tcolorbox}
\usepackage{zref-clever}
\usepackage{hyperref}

\tcbset{%
  zlabel/.style={phantom={\zlabel{#1}}},%
  step and zlabel/.style 2 args={phantom={\refstepcounter{#1}\zlabel{#2}}},%
  phantomzlabel/.style={phantom={\ifdefined\phantomsection\phantomsection\fi\zlabel{#1}}},%
  zlabel type/.code={\zcsetup{reftype={#1}}},%
}

\ExplSyntaxOn
\makeatletter
\tcbset{
  new/.cd,
  zref-clever~countertype/.code={
    \exp_args:Ne\zcsetup{countertype={tcb@cnt@\tcb@new@boxname={#1}}}
  },
}
\makeatother
\ExplSyntaxOff

\newtcolorbox[auto counter,number within=section,zref-clever countertype=example]{pabox}[2][]{%
  colback=red!5!white,colframe=red!75!black,fonttitle=\bfseries,
  title=Examp.~\thetcbcounter: #2,#1}

% A custom type
\newtcolorbox[auto counter,number within=section,zref-clever countertype=myexample]{mypabox}[2][]{%
  colback=red!5!white,colframe=red!75!black,fonttitle=\bfseries,
  title=Examp.~\thetcbcounter: #2,#1}
\zcRefTypeSetup{myexample}{
  % that which crefname/Crefname would provide:
  Name-sg=Myexample,
  name-sg=myexample,
  Name-pl=Myexamples,
  name-pl=myexamples,
  % and more, e.g.:
  reffont=\bfseries,
  refbounds={(,,,)},
  refbounds-first-sg={,(,),},
}

\begin{document}

\section{Foo bar}
\zlabel{sec:section-1}

\begin{pabox}[zlabel={myautocounter}]{Title with number}
  This box is automatically numbered with \zcref{myautocounter} on
  \zcpageref{myautocounter}. Inside the box, the \thetcbcounter\ can also be
  referenced by |\thetcbcounter|.  The real counter name is
  \texttt{\tcbcounter}.
\end{pabox}

\begin{pabox}[zlabel={typedlabel},zlabel type=solution]{Title with number}
  This box is automatically numbered with \zcref{typedlabel} on
  \zcpageref{typedlabel}. Inside the box, the \thetcbcounter\ can also be
  referenced by |\thetcbcounter|.  The real counter name is
  \texttt{\tcbcounter}.
\end{pabox}

\begin{mypabox}[zlabel={customtype}]{Title with number}
  This box is automatically numbered with \zcref{customtype} on
  \zcpageref{customtype}. Inside the box, the \thetcbcounter\ can also be
  referenced by |\thetcbcounter|.  The real counter name is
  \texttt{\tcbcounter}.
\end{mypabox}

\end{document}
```

WDYT?

**Comment by T-F-S** [comment]
That sounds very promising. I will look deeper into it the next days, but if nothing critical comes up, this seems to be a valuable new feature for `tcolorbox`.

**Comment by T-F-S** [comment]
Is there a reasonable application case where `\label` *and* `\zlabel` are needed in parallel? My answer would be no, but, maybe, I missed something.

I imagine a switch like `label is label` and `label is zlabel` to change the meaning of the `label` option for `tcolorbox`. This would also change `step and label` and `phantomlabel` to use either `\label` or `\zlabel`. This would give the advantage that the theorem enviroments of `tcolorbox` are automatically `zref-clever` compatible without further hacks.

The proposed initialization option `zref-clever countertype` is not applicable for cases where a shared counter is used. But, it would be possible to have such an option as regular option for `tcolorbox` which is similar to the *shared counter* documentation of `zref-clever`.

I wonder, if a regular option `zref-clever countertype` does anything other than the proposed option `zlabel type`? I am not familiar enough with  `zref-clever`, but I have the impression that `zlabel type` can do anything needed (?).

Finally, with a switch like `label is zlabel`, `label type` could switch its meaning to `zlabel type`.

**Comment by gusbrs** [comment]
Hi Thomas, thank you for looking into this.

> Is there a reasonable application case where `\label` _and_ `\zlabel` are needed in parallel? My answer would be no, but, maybe, I missed something.
> 
> I imagine a switch like `label is label` and `label is zlabel` to change the meaning of the `label` option for `tcolorbox`. This would also change `step and label` and `phantomlabel` to use either `\label` or `\zlabel`. This would give the advantage that the theorem enviroments of `tcolorbox` are automatically `zref-clever` compatible without further hacks.

I'd tend to suggest keeping things separate. My reasoning being the following.

The referencing systems are independent, and they can be used in parallel, including with the same labels. Also there are cases for which there's no support for `\zlabel`. The known, and most relevant one, are `amsmath` display math environments. This is more or less "official" (https://tex.stackexchange.com/a/402297), and considering the stability requirements of `amsmath` there's little hope of getting support eventually. `zref-clever` handles this by taking a "ride" on `\label`, but `\label` must be used inside the environments (more or less this procedure https://tex.stackexchange.com/a/618438). So, this is one case where one not only can but actually needs to use them in parallel. Furthermore, if one would like to convert a document from `cleveref` to `zref-clever`, `\zlabel` is not a drop-in replacement for `\label` (`cleveref`'s version of it), because it does not take an optional argument. So the traditional trick to "convert" footnotes to endnotes does not apply here, and we cannot really just `\let\label\zlabel`.

That said, I do get the appeal of your proposal too. Not only for the user convenience of "just switching", but also from the package's UI perspective. And, admittedly, the above issues would not affect `tcolorbox` references, and a switch would indeed work well. All in all, though I have the mentioned inclination, I have no strong preference, and would be very happy with whatever you think is best in this regard.

> The proposed initialization option `zref-clever countertype` is not applicable for cases where a shared counter is used. But, it would be possible to have such an option as regular option for `tcolorbox` which is similar to the _shared counter_ documentation of `zref-clever`.

I'm not sure I understand your concern here. As far as I do, I'd say the option is redundant when used with `use counter from`, which is indeed not desirable. Perhaps, it could be inhibited, similar to what is done with `number format` and `number within` (I haven't checked what is done about them, but it'd be the same problem, as far as I understand).

> I wonder, if a regular option `zref-clever countertype` does anything other than the proposed option `zlabel type`? I am not familiar enough with `zref-clever`, but I have the impression that `zlabel type` can do anything needed (?).

Indeed, they serve the same purpose, but they work differently. And in this regard, `zref-clever` is different from `cleveref`. Among the many qualities of `cleveref` I would not list the inclusion of an optional argument to `\label`. True, `cleveref` could hardly be able to get by without redefining `\label`, but the optional argument was avoidable, and its inclusion is the source of a good number of compatibility issues and headaches spread across the ecosystem. 

With that in mind, `zref-clever` does not touch `\zlabel` at all, which remains as provided by `zref`. And it works by setting some variables which are available with appropriate values when `\zlabel` is set. `countertype` and `reftype` both affect the `zc@type` property, which defines the reference type, but in different ways. `countertype` sets a property list associating a counter with its type. When the label is set, if a counter is found in the list, the type is retrieved, otherwise, the counter itself is used as the type. `reftype` is a brute force override, intended for local use only, and sets the type regardless of the counter. So, if you set `reftype`, the zlabel set by a tcolorbox will get the expected type, but so will any other `\zlabel`s set within the scope. Now, suppose you have a box with a table within it, and you'd like to refer to the table rather than the box. That wouldn't work well with `reftype` since the table would be stored with the type specified by `reftype`, whereas if we just set the `countertype` for the box's counter, we'd get the expected types for both our labels.

> Finally, with a switch like `label is zlabel`, `label type` could switch its meaning to `zlabel type`.

Adding to what was said above, the optional argument of `\label` in `cleveref` does not translate well into `zref-clever`'s structure. They are similar but not equal. The switch in this case might produce undesirable results.

**Comment by gusbrs** [comment]
> > The proposed initialization option `zref-clever countertype` is not applicable for cases where a shared counter is used. But, it would be possible to have such an option as regular option for `tcolorbox` which is similar to the _shared counter_ documentation of `zref-clever`.
> 
> I'm not sure I understand your concern here. As far as I do, I'd say the option is redundant when used with `use counter from`, which is indeed not desirable. Perhaps, it could be inhibited, similar to what is done with `number format` and `number within` (I haven't checked what is done about them, but it'd be the same problem, as far as I understand).

My bad, I seem to have forgotten my own documentation. ;-)

Revising my statement, I think the option would apply to a new box with `use counter from` just as well. `zref-clever`'s `countertype` option would apply to the environment locally. Or am I missing something?

**Comment by gusbrs** [comment]
> The application of `countertype` is/was more tricky than it looks like, because it has to be placed after the local counter is set up and before `\zlabel` is used. I have some test code now which puts it inside `\tcb@set@label`. The test code is not presentable and currently my time is running out. But, I intend to continue tomorrow or end of this week.

Ah, you meant that the way I proposed to provide the `zref-clever countertype` is simplistic and does not handle ways other than an `auto counter` to set the box's counter. And you are right, of course. I had assumed the only situation a user would need the option was when `auto counter` was used, since the user then never gets to know the name of the counter (unless they have read the docs, which we cannot take for granted). But, for starters, there's also `use counter from`, which can refer to a counter of a box initialized with `auto counter`. Besides, if the option exists, why shouldn't it apply to `use counter` and `user counter*`? Good point.

I took a quick look now, and couldn't figure out a good way to generalize it. You're probably better positioned to know the right spot to intervene for this purpose, but I will try further tomorrow as well and report back if I come up with any ideas.

**Comment by gusbrs** [comment]
Hi Thomas,

I've been trying things out this morning and still wasn't able to find a good solution, but I had some further ideas, which I might as well share.

The reason I created the `reftype` option as it is was so that users didn't have to look up the name of the counter, as `countertype` requires. That is its only "advantage", and it does have some disadvantages mentioned above. But thinking this through in the context of `tcolorbox`, and since we know the name of the counter (`\tcbcounter`), there's really no reason to ever use `reftype`. And, by now, I think we should be using only `countertype` here, given we have `\tcbcounter`.

Also, that given, and contrary to what I had initially suggested for the `zref-clever countertype` option, I think this should be just an option to the box itself, and set in the `[<default>]` argument of `\newtcolorbox`, instead of in the `[<init>]` argument. In other words, just use `\tcbset{zlabel~type/.code={\exp_args:Ne\zcsetup{countertype={\tcbcounter={#1}}}}}` (of course, or something equivalent, pending on your decision regarding the switch).

True, as you pointed out, that still doesn't work, because it comes at the wrong timing. As you noted, it must come after `\tcbcounter` is set, and before the label is set. So far I wasn't able to find the right spot to do it. I'll keep trying this afternoon, but I admit I'm already out of my depth in `tcolorbox`'s internals, so I'm not sure I'll be able to come through on this one.

**Comment by T-F-S** [comment]
> True, as you pointed out, that still doesn't work, because it comes at the wrong timing. As you noted, it must come after `\tcbcounter` is set, and before the label is set. So far I wasn't able to find the right spot to do it. I'll keep trying this afternoon, but I admit I'm already out of my depth in `tcolorbox`'s internals, so I'm not sure I'll be able to come through on this one.

I chose the save way and put it inside `tcolorbox`'s `label` mechanism. The only drawback is that the countertype value needs to be stored and resetted for inner boxes (this is another story, but it enlarges the footprint of a `tcolorbox`). The reftype is much easier, because code can be executed directly, but as you pointed out, it punches through to every zlabel in the environment.

I renamed `zlabel type` to `zref-cleber reftype` for consistency. Now, my code is the following:

```latex
\documentclass{article}

\usepackage{tcolorbox}
\tcbuselibrary{theorems}

\usepackage{zref-titleref}
\usepackage{zref-clever}
\usepackage{hyperref}


\ExplSyntaxOn
\makeatletter

\tl_new:N \l__tcobox_kv_labeltype_tl
\tl_new:N \l__tcobox_kv_zc_countertype_tl


\tcbset{
  label          /.style = { phantom={\__tcobox_label:n{#1}} },
  phantomlabel   /.style = { phantom={\cs_if_exist_use:N\phantomsection\__tcobox_label:n{#1}} },
  step~and~label /.style~n~args = { 2 }{ phantom={\refstepcounter{#1}\__tcobox_label:n{#2}} },
  nameref        /.code =
    {
      \tcb@gettitle{#1}
      \cs_if_exist:NT \zref@titleref@setcurrent { \zref@titleref@setcurrent{#1} }
    },

  label~type    / .store~in = \l__tcobox_kv_labeltype_tl,
  no~label~type / .style    = {label~type=},

  zref-clever~countertype    /.store~in = \l__tcobox_kv_zc_countertype_tl,
  no~zref-clever~countertype /.style    = { zref-clever~countertype=\q_nil },
  zref-clever~reftype /.code = { \zcsetup{reftype={#1}} },

}


\cs_new:Npn \__tcobox_label_label:n #1
  {
    \tl_if_empty:nTF \l__tcobox_kv_labeltype_tl
      {
        \label{#1}
      }
      {
        \exp_last_unbraced:NNV \label[\l__tcobox_kv_labeltype_tl]{#1}
      }
  }


\cs_new:Npn \__tcobox_zcsetup_countertype:nn #1#2
  {
    \zcsetup{countertype={#1={#2}}}
  }


\cs_new:Npn \__tcobox_label_zlabel:n #1
  {
    \quark_if_nil:NF \l__tcobox_kv_zc_countertype_tl
      {
        \cs_if_exist:NT \tcbcounter
          {
            \exp_args:NVV \__tcobox_zcsetup_countertype:nn \tcbcounter \l__tcobox_kv_zc_countertype_tl
          }
      }
    \zlabel{#1}
  }


\tcbset{
  label~is~label/.code=
    {
      \cs_set_eq:NN \__tcobox_label:n \__tcobox_label_label:n
    },
  label~is~zlabel/.code=
    {
      \cs_set_eq:NN \__tcobox_label:n \__tcobox_label_zlabel:n
    },
}


\tcbset{
  label~is~label,% ***** no reset *****
  no~zref-clever~countertype,% ********** reset *********
}


\makeatother
\ExplSyntaxOff



\newtcolorbox[auto counter,number within=section]{pabox}[2][]{%
  colback=red!5!white,colframe=red!75!black,fonttitle=\bfseries,
  zref-clever countertype=example,
  title=Examp.~\thetcbcounter: #2,#1}

\newcounter{extracounter}


% A custom type
\newtcolorbox[%
  use counter=extracounter,
  number within=section,
  ]{mypabox}[2][]{%
  colback=red!5!white,colframe=red!75!black,fonttitle=\bfseries,
  zref-clever countertype=myexample,
  title=Myexample~\thetcbcounter: #2,#1}

\zcRefTypeSetup{myexample}{
  % that which crefname/Crefname would provide:
  Name-sg=Myexample,
  name-sg=myexample,
  Name-pl=Myexamples,
  name-pl=myexamples,
  % and more, e.g.:
  reffont=\bfseries,
  refbounds={(,,,)},
  refbounds-first-sg={,(,),},
}

\newtcbtheorem[number within=section]{mytheo}{My Theorem}%
  {colback=green!5,colframe=green!35!black,fonttitle=\bfseries,
   zref-clever countertype=mytheorem}{th}

\zcRefTypeSetup{mytheorem}{
  % that which crefname/Crefname would provide:
  Name-sg=Mytheorem,
  name-sg=mytheorem,
  Name-pl=Mytheorems,
  name-pl=Mytheorems,
  % and more, e.g.:
  %reffont=\bfseries,
  %refbounds={(,,,)},
  %refbounds-first-sg={,(,),},
}


\begin{document}

\tcbset{label is zlabel}

\section{Foo bar}
\zlabel{sec:section-1}

\begin{pabox}[label={myautocounter}]{Title with number}
  This box is automatically numbered with \zcref{myautocounter} on
  \zcpageref{myautocounter}. Inside the box, the \thetcbcounter\ can also be
  referenced by |\thetcbcounter|.  The real counter name is
  \texttt{\tcbcounter}.
\end{pabox}

\begin{pabox}[label={typedlabel},zref-clever reftype=solution]{Title with number}
  This box is automatically numbered with \zcref{typedlabel} on
  \zcpageref{typedlabel}. Inside the box, the \thetcbcounter\ can also be
  referenced by |\thetcbcounter|.  The real counter name is
  \texttt{\tcbcounter}.
\end{pabox}

\begin{mypabox}[label={customtype}]{Title with number}
  This box is automatically numbered with \zcref{customtype} on
  \zcpageref{customtype}. Inside the box, the \thetcbcounter\ can also be
  referenced by |\thetcbcounter|.  The real counter name is
  \texttt{\tcbcounter}.
\end{mypabox}

\begin{pabox}[label={xtypedlabel}]{Title with number}
  This box is automatically numbered with \zcref{xtypedlabel} on
  \zcpageref{xtypedlabel}. Inside the box, the \thetcbcounter\ can also be
  referenced by |\thetcbcounter|.  The real counter name is
  \texttt{\tcbcounter}.
\end{pabox}

\newcounter{othercounter}

\begin{pabox}[label={ytypedlabel},zref-clever reftype=myexample]{Title with number}
  This box is automatically numbered with \zcref{ytypedlabel} on
  \zcpageref{ytypedlabel}. Inside the box, the \thetcbcounter\ can also be
  referenced by |\thetcbcounter|. The real counter name is
  \texttt{\tcbcounter}.
  \stepcounter{othercounter}\zlabel{other1}
  Other reference is \zcref{other1}.
\end{pabox}


\begin{pabox}[label={ztypedlabel},zref-clever countertype=myexample]{Title with number}
  This box is automatically numbered with \zcref{ztypedlabel} on
  \zcpageref{ztypedlabel}. Inside the box, the \thetcbcounter\ can also be
  referenced by |\thetcbcounter|.  The real counter name is
  \texttt{\tcbcounter}.
  \stepcounter{othercounter}\zlabel{other2}
  Other reference is \zcref{other2}.
\end{pabox}


\begin{tcolorbox}[label={dummy},zref-clever countertype=myexample]
Some box
\end{tcolorbox}


\begin{mytheo}{This is my title}{theoexample}
  This is the text of the theorem. The counter is automatically assigned and,
  in this example, prefixed with the section number. This theorem is numbered with
  \zcref[cap]{th:theoexample}, it is given on \zcpageref{th:theoexample},
  and it is titled ``\ztitleref{th:theoexample}''.
\end{mytheo}


\end{document}
```

Next, I will integrate this into `tcolorbox`. If no problems appear, it will be part of the next version.

**Comment by T-F-S** [comment]
> The reason I created the `reftype` option as it is was so that users didn't have to look up the name of the counter, as `countertype` requires. That is its only "advantage", and it does have some disadvantages mentioned above. But thinking this through in the context of `tcolorbox`, and since we know the name of the counter (`\tcbcounter`), there's really no reason to ever use `reftype`. And, by now, I think we should be using only `countertype` here, given we have `\tcbcounter`.

We were working in parallel and I missed your thoughts about `reftype`.

I agree to drop the `zref-clever reftype` option. It also would be difficult to explain a user the difference to `zref-clever countertype` and what to take.

**Comment by T-F-S** [comment]
I uploaded a new preview version with the support code: https://github.com/T-F-S/tcolorbox/commit/54c1dd84fdd601368edcf52e79b30122cd80185e

**Comment by gusbrs** [comment]
@T-F-S , I've taken a look at things, and it looks great. Thank you very much. I've yet to test it, but I do have one suggestion at this point. Since you've chosen the `label is label/zlabel` switch, and also to drop the `reftype` option, it might make sense to incorporate `zref-clever countertype` into `label type`, so that the switch is "complete". `countertype` is not *identical* to the use `tcolorbox` makes of the optional argument of `\label`, because with the latter, only the label set by the `label` option has it's type changed, whereas with `countertype`, as long as no one has issued a `\refstepcounter`, a plain label issued by the user inside the box will also be affected. But I think that their meaning is close enough, and the difference is arguably desirable, to grant their merging into `label type`. WDYT?

**Comment by gusbrs** [comment]
Hi Thomas, I've done some testing now, and I can gladly report that it is looking good also in that regard. The only corner case I've found is that of a unnubered box. Consider the document below:

```tex
\documentclass{article}

\usepackage{tcolorbox}
\tcbuselibrary{theorems}

\usepackage{zref-titleref}
\usepackage{zref-clever}
\usepackage{hyperref}

\usepackage{cleveref}

\ExplSyntaxOn
\makeatletter

\tl_new:N \l__tcobox_kv_labeltype_tl
\tl_new:N \l__tcobox_kv_zc_countertype_tl


\tcbset{
  label          /.style = { phantom={\__tcobox_label:n{#1}} },
  phantomlabel   /.style = { phantom={\cs_if_exist_use:N\phantomsection\__tcobox_label:n{#1}} },
  step~and~label /.style~n~args = { 2 }{ phantom={\refstepcounter{#1}\__tcobox_label:n{#2}} },
  nameref        /.code =
    {
      \tcb@gettitle{#1}
      \cs_if_exist:NT \zref@titleref@setcurrent { \zref@titleref@setcurrent{#1} }
    },

  label~type    / .store~in = \l__tcobox_kv_labeltype_tl,
  no~label~type / .style    = {label~type=},

  zref-clever~countertype    /.store~in = \l__tcobox_kv_zc_countertype_tl,
  no~zref-clever~countertype /.style    = { zref-clever~countertype=\q_nil },
}


\cs_new:Npn \__tcobox_label_label:n #1
  {
    \tl_if_empty:nTF \l__tcobox_kv_labeltype_tl
      {
        \label{#1}
      }
      {
        \exp_last_unbraced:NNV \label[\l__tcobox_kv_labeltype_tl]{#1}
      }
  }


\cs_new:Npn \__tcobox_zcsetup_countertype:nn #1#2
  {
    \zcsetup{countertype={#1={#2}}}
  }


\cs_new:Npn \__tcobox_label_zlabel:n #1
  {
    \quark_if_nil:NF \l__tcobox_kv_zc_countertype_tl
      {
        \cs_if_exist:NT \tcbcounter
          {
            \exp_args:NVV \__tcobox_zcsetup_countertype:nn \tcbcounter \l__tcobox_kv_zc_countertype_tl
          }
      }
    \zlabel{#1}
  }


\tcbset{
  label~is~label/.code=
    {
      \cs_set_eq:NN \__tcobox_label:n \__tcobox_label_label:n
    },
  label~is~zlabel/.code=
    {
      \cs_set_eq:NN \__tcobox_label:n \__tcobox_label_zlabel:n
    },
}


\tcbset{
  label~is~label,% ***** no reset *****
  no~zref-clever~countertype,% ********** reset *********
}


\makeatother
\ExplSyntaxOff

\begin{document}

\section{Foo bar}
\zlabel{sec:section-1}

\tcbset{label is zlabel}

\begin{tcolorbox}[label={dummy},zref-clever countertype=equation]
Some box \zcref{dummy}.
\end{tcolorbox}

\tcbset{label is label}

\begin{tcolorbox}[label={dummy},label type=equation]
Some box \cref{dummy}.
\end{tcolorbox}

\end{document}
```
The `\zcref` reference comes out as "section 1", while the `\cref` reference as "eq. (1)". I'm tempted to say the first one is more "correct", since `\refstepcounter` was not called for the box. But it is a little counterintuitive that the type was manually specified and does not take effect, even if this input should probably be regarded as "user error" to start with. Either way, the results are somewhat inconsistent between the two "label backends" as things stand. I'm not sure this difference should be resolved, since it is a fact that `countertype` requires `\tcbcounter` to be set, while the optional argument of `\label` does not. Still, I thought it worth reporting for your consideration.

**Comment by gusbrs** [comment]
Mhm, I caught a small typo with some nasty consequences. Consider the document:

```tex
\documentclass{article}

\usepackage{tcolorbox}
\tcbuselibrary{theorems}

\usepackage{hyperref}

\ExplSyntaxOn
\makeatletter

\tl_new:N \l__tcobox_kv_labeltype_tl
\tl_new:N \l__tcobox_kv_zc_countertype_tl

\tcbset{
  label          /.style = { phantom={\__tcobox_label:n{#1}} },
  phantomlabel   /.style = { phantom={\cs_if_exist_use:N\phantomsection\__tcobox_label:n{#1}} },
  step~and~label /.style~n~args = { 2 }{ phantom={\refstepcounter{#1}\__tcobox_label:n{#2}} },
  nameref        /.code =
    {
      \tcb@gettitle{#1}
      \cs_if_exist:NT \zref@titleref@setcurrent { \zref@titleref@setcurrent{#1} }
    },

  label~type    / .store~in = \l__tcobox_kv_labeltype_tl,
  no~label~type / .style    = {label~type=},

  zref-clever~countertype    /.store~in = \l__tcobox_kv_zc_countertype_tl,
  no~zref-clever~countertype /.style    = { zref-clever~countertype=\q_nil },
}


\cs_new:Npn \__tcobox_label_label:n #1
  {
    \tl_if_empty:nTF \l__tcobox_kv_labeltype_tl
      {
        \label{#1}
      }
      {
        \exp_last_unbraced:NNV \label[\l__tcobox_kv_labeltype_tl]{#1}
      }
  }


\cs_new:Npn \__tcobox_zcsetup_countertype:nn #1#2
  {
    \zcsetup{countertype={#1={#2}}}
  }


\cs_new:Npn \__tcobox_label_zlabel:n #1
  {
    \quark_if_nil:NF \l__tcobox_kv_zc_countertype_tl
      {
        \cs_if_exist:NT \tcbcounter
          {
            \exp_args:NVV \__tcobox_zcsetup_countertype:nn \tcbcounter \l__tcobox_kv_zc_countertype_tl
          }
      }
    \zlabel{#1}
  }


\tcbset{
  label~is~label/.code=
    {
      \cs_set_eq:NN \__tcobox_label:n \__tcobox_label_label:n
    },
  label~is~zlabel/.code=
    {
      \cs_set_eq:NN \__tcobox_label:n \__tcobox_label_zlabel:n
    },
}


\tcbset{
  label~is~label,% ***** no reset *****
  no~zref-clever~countertype,% ********** reset *********
}


\makeatother
\ExplSyntaxOff


\newtcolorbox[auto counter,number within=section]{pabox}[2][]{%
  colback=red!5!white,colframe=red!75!black,fonttitle=\bfseries,
  title=Examp.~\thetcbcounter: #2,#1}

\begin{document}

\section{Foo bar}
\label{sec:section-1}

\begin{pabox}[label={myautocounter}]{Title with number}
  This box is automatically numbered with \ref{myautocounter} on
  \pageref{myautocounter}. Inside the box, the \thetcbcounter\ can also be
  referenced by |\thetcbcounter|.  The real counter name is
  \texttt{\tcbcounter}.
\end{pabox}

\end{document}
```
This will result in undefined references for `myautocounter`, and inspection of the `.aux` file will find the label as `\newlabel{[}{{1.1}{1}{Foo bar}{tcb@cnt@pabox.1.1}{}}` meaning the optional argument branch in `\__tcobox_label_label:n` is being chosen even though `label type` was never specified. As far as I can tell, you should be using `\tl_if_empty:NTF` instead of `\tl_if_empty:nTF` for the test.

**Comment by gusbrs** [comment]
And I have one further suggestion to make, namely to protect the call to `\zcsetup` inside `\__tcobox_zcsetup_countertype:nn` with `\cs_if_exist:NT \zcsetup`. It is not possible to test whether `\label` has an optional argument in the case of `cleveref`, but here it is trivial to test if `\zcsetup` exists, and we get a compiling document even if the user commits the mistake of setting `zref-clever countertype` (or `label type`, pending my suggestion) in the absence of `zref-clever`. Perhaps we could even extend this logic to `\zlabel` itself, since it is not available from the kernel, as `\label` is.

**Comment by T-F-S** [comment]
> @T-F-S , I've taken a look at things, and it looks great. Thank you very much. I've yet to test it, but I do have one suggestion at this point. Since you've chosen the `label is label/zlabel` switch, and also to drop the `reftype` option, it might make sense to incorporate `zref-clever countertype` into `label type`, so that the switch is "complete". `countertype` is not _identical_ to the use `tcolorbox` makes of the optional argument of `\label`, because with the latter, only the label set by the `label` option has it's type changed, whereas with `countertype`, as long as no one has issued a `\refstepcounter`, a plain label issued by the user inside the box will also be affected. But I think that their meaning is close enough, and the difference is arguably desirable, to grant their merging into `label type`. WDYT?

I had the same thought, but I tend to stay at two distinct options.
- `\label`/`\zlabel` are different, but both set just labels. Switching generates no errors, just ?? signs.
- `label type` for cleveref and `zref-clever countertype` under one name would produce errors, if the user switches to the other system.
- `label type` is nullified by an empty string while `zref-clever countertype` is nullified by `\q_nil`, because according to your documentation an empty string is a valid option. Arguably, the implementation for `label type` could be adapted.
- IMHO a user could become puzzled, if both options are melted together. Currently, my documention draft is the following:
![grafik](https://user-images.githubusercontent.com/7982989/217797729-98eec656-f3fd-4a36-9fb4-6d4335bc611d.png)

**Comment by T-F-S** [comment]
> This will result in undefined references for `myautocounter`, and inspection of the `.aux` file will find the label as `\newlabel{[}{{1.1}{1}{Foo bar}{tcb@cnt@pabox.1.1}{}}` meaning the optional argument branch in `\__tcobox_label_label:n` is being chosen even though `label type` was never specified. As far as I can tell, you should be using `\tl_if_empty:NTF` instead of `\tl_if_empty:nTF` for the test.

You are completely right: This is a bug. I'm happy that you found it before an official release.

**Comment by T-F-S** [comment]
> Regarding your objections, I may be missing something, but why would it produce errors? The sets of predefined types for both `cleveref` and `zref-clever` are not identical, but they mostly overlap, and both `cleveref` and `zref-clever` fail gracefully if they meet a type they don't know: we get `??` for the name and a compilation warning (not an error). The only cases where errors could occur would be if: a) a `\label` with optional argument is called, but `cleveref` is not present, but that's already the case currently; b) if `\zcsetup` is called and `zref-clever` is not present, but that's easy to guard against, with `\cs_if_exist:NT` as suggested, and as the documentation draft seems to hint has been incorporated.

I think I mixed things up with errors from a missing `\zcsetup` before protection with `\cs_if_exist:NT`.

> As to `countertype` accepting an empty value to "unset" an association from the list, this is true. But `tcolorbox` could either not support this minor feature (which is mostly aimed at undoing some global setting of the option, while all settings here are local, besides the same end result could always be achieved with `zref-clever countertype=\tcbcounter`) and test for emptiness of `\l__tcobox_kv_zc_countertype_tl` / `\l__tcobox_kv_labeltype_tl`, or it could test `label type` for `\q_nil` and, if it is, the standard `\label` with no optional argument is called and the value of the option is discarded anyway.

OK, I finally melt `zref-clever countertype` into `label type` and use `\q_nil` to nullify.
This has one additional advantage: It removes one variable from the `reset` list of `tcolorbox`.

**Comment by gusbrs** [comment]
Hi @T-F-S , I have taken a look at the new commits now, done some further testing and, to the best of my knowledge, everything is looking great. So I have nothing further to add, except: Vielen Dank! :-)

I'll be looking forward to the next release. And, of course, if you still need anything from me in this regard, I remain at your disposal. Don't hesitate to let me know.

**Comment by T-F-S** [comment]
Implemented with https://github.com/T-F-S/tcolorbox/releases/tag/v6.0.0

**Comment by gusbrs** [comment]
Once again, thank you very much!
